Updates to Dimensionality reduction notebook

jashapiro opened the pull request:
This update to the dimensionality reduction notebooks substitutes in the 10X Hodgkins lymphoma data set. 

New content added includes:
- demonstrating how to read in Cell Ranger output #451 
- emptydrops filtering


Modifications:
- dimensionality reduction is simplified, focusing on using the SingleCellExperiment functions for analysis.
- Identifying marker genes has been removed, with the idea that it will be added in a separate notebook that focuses on clustering methods (as described in #455)
- I did not eliminate the experimentation completely... I decided that having that break was worth keeping, but I did simplify the function, much as I hated to lose the bare ggplot...

Reviewers should focus on the overall flow here first: does everything still make sense as presented? If that seems good, more specific comments are welcome! Are there some additional references that I should add? I am not sure I give enough credit to the Orchestrating Single Cell book, though I expect more of that in the next notebook!

closes #451 

Note: I'm also expecting CI to fail here at least once...

jaclyn-taroni commented on scRNA-seq/04-dimension_reduction_scRNA.Rmd:
```diff
@@ -1,12 +1,23 @@
+- Read Cell Ranger data into R 
+- Filter post-quantification cells using `emptyDrops()`
+- Apply dimensionality reduction methods to single cell data
+- Visualize reduced dimensionality data
```
I haven't read further, but I assume that this is what we're doing and I think being a bit more specific could be helpful:

```suggestion
- Visualize samples in reduced dimensional space
```

jaclyn-taroni commented on scRNA-seq/04-dimension_reduction_scRNA.Rmd:
```diff
@@ -31,225 +44,255 @@ library(magrittr)
+The `DropletUtils::read10xCounts()` function takes this directory and reads in the data from these three files, assembling the `SingleCellExperiment` object we have worked with before. 
+
+Alternatively, we could use the `HDF5` format file that Cell Ranger outputs as a file with the `.h5` extension, which contains the same data. 
+For whatever reason, the way you read the data affects how it is stored in R, with the net result that reading from the directory results in smaller objects in R, so that is what we will do here. 
```
```suggestion
For whatever reason, the way you read the data affects how it is stored in R. 
Reading from the directory results in smaller objects in R, so that is what we will do here. 
```

jaclyn-taroni commented on scRNA-seq/04-dimension_reduction_scRNA.Rmd:
````diff
@@ -31,225 +44,255 @@ library(magrittr)
-Using these QC data, we can now create a subset of the `SingleCellExperiment` object, keeping only the columns (samples) that pass our QC thresholds:
+```{r mito_percent_plot, live = TRUE}
+# Plot the mitochondrial percents stored in `filtered_sce`
+ggplot( mapping = aes(x = filtered_sce$subsets_mito_percent)) +
````
```suggestion
ggplot(mapping = aes(x = filtered_sce$subsets_mito_percent)) +
```

jaclyn-taroni commented on scRNA-seq/04-dimension_reduction_scRNA.Rmd:
````diff
@@ -31,225 +44,255 @@ library(magrittr)
 ```
 
+
 We can also filter by features (genes in our case) using `scater::addPerFeatureQC()` which will computer the number of samples where each gene is detected and the mean count across all genes. 
````
I mean computer isn't wrong

```suggestion
We can also filter by features (genes in our case) using `scater::addPerFeatureQC()` which will compute the number of samples where each gene is detected and the mean count across all genes. 
```

jaclyn-taroni commented on scRNA-seq/04-dimension_reduction_scRNA.Rmd:
```diff
@@ -31,225 +44,255 @@ library(magrittr)
 ### Normalize
 
-Now we will perform the same normalization steps we did with the Smart-seq2 data, using `scran::computeSumFactors()` and `scater::logNormCounts()`
+Now we will perform the same normalization steps we did with a previous dataset, using `scran::computeSumFactors()` and `scater::logNormCounts()`
```
Reminder that we set the seed earlier?

jaclyn-taroni commented on scRNA-seq/04-dimension_reduction_scRNA.Rmd:
```diff
@@ -31,225 +44,255 @@ library(magrittr)
+
+### Principal Components Analysis
 
 The first dimensionality reduction algorithm we will use is PCA, which we have already used!
 The idea of PCA is that it finds the axes which explain the greatest amount of variance in the data.
```
That will not be true during this training as far as I know, so would port over some of the text in: https://github.com/AlexsLemonade/training-modules/blob/0388bf1c7c04fbbc5dd3ff64f8eae98b2472826c/RNA-seq/03-gastric_cancer_exploratory.Rmd#L112

jaclyn-taroni commented on scRNA-seq/04-dimension_reduction_scRNA.Rmd:
````diff
@@ -31,225 +44,255 @@ library(magrittr)
-```{r load_normalized}
-tm_normalized <- readr::read_rds(tm_normalized_file)
+It looks like there may be a few clusters there, though they are not very distinct.
+Sometimes looking at higher dimensions can show more (or different) structure, so lets have a look there too.
````
```suggestion
Sometimes looking at higher dimensions can show more (or different) structure, so let's have a look there too.
```

jashapiro replied:
I make this mistake too much. Thank you for always catching it!

jaclyn-taroni commented on scRNA-seq/04-dimension_reduction_scRNA.Rmd:
````diff
@@ -31,225 +44,255 @@ library(magrittr)
 ```
 
-How many cells are in this data? How many samples are there? How many cells from each tissue?
+#### Storing PCA results with the data
````
In terms of flow, it's not clear to me why we wouldn't use `runPCA()` first? Unless plotting using `reducedDim()` is a pain. I would think we want to encourage folks to default to saving the results as part of the object.

jashapiro replied:
This makes sense. I was going back and forth about how much to rely on the object vs. tidyverse standards, and I have been leaning more and more toward just using the object... I worry that it requires more abstract thinking and I worry that people will build odd mental models of how it is working, but the convenience level (and consistency benefits) are pretty high.

jaclyn-taroni commented on scRNA-seq/04-dimension_reduction_scRNA.Rmd:
```diff
@@ -263,87 +306,61 @@ There is some ongoing debate about which of these two techniques is superior, an
+We will come back to this question later!
 
-## UMAP experiments
+### UMAP experiments
```
No one is going to be surprised to hear this but I do enjoy this part a lot.

jaclyn-taroni commented on scRNA-seq/04-dimension_reduction_scRNA.Rmd:
```diff
@@ -263,87 +306,61 @@ There is some ongoing debate about which of these two techniques is superior, an
 For even more parameters, you can look at the underlying implementation code that `calculateUMAP()` uses, which is the function `uwot::umap()`
 
-In order to make our experimentation easier, we will create a *function* that allows us to rerun the same code chunk easily, but create an argument that allows us to change one variable: the n_neighbors variable. 
+In order to make our experimentation easier, we will create a *function* that allows us to rerun the same code easily, but create an argument that allows us to change one variable: the n_neighbors variable.
```
```suggestion
In order to make our experimentation easier, we will create a *function* that allows us to rerun the same code easily, but create an argument that allows us to change one variable: the `n_neighbors` variable.
```

jaclyn-taroni commented on scRNA-seq/04-dimension_reduction_scRNA.Rmd:
```diff
@@ -381,44 +398,49 @@ UMAP_plot_wrapper(nn_param = 5)
 
+
+
 ### t-SNE comparison
```
Hm I could talk myself into removing this if we're concerned about time.

jashapiro replied:
It's just a little snack at the end. And I have plans to coming back to it later...

jaclyn-taroni commented on scRNA-seq/04-dimension_reduction_scRNA.Rmd:
```diff
@@ -31,225 +44,255 @@ library(magrittr)
+The `DropletUtils` package that we used to read in the 10X data has a number of other useful features. 
+One is the `emptyDrops()` function, which uses the overall gene expression patterns in the sample to identify droplets that are likely to not contain an intact cell, but may simply have contained loose ambient RNA released during cell separation. 
+
+This method, described in [Lun *et al.* (2019)](https://doi.org/10.1186/s13059-019-1662-y), uses the droplets with very low UMI counts to estimate the "ambient" expression pattern, then scores the remaining cells based how much they deviate from that pattern, assigning a P value based on the null hypothesis that the droplet's expression matches the ambient expression pattern.
```
Might break up this sentence

jaclyn-taroni reviewed (commented):
I would say the overall flow of this looks good! The two most substantive comments about flow I had were:

* Do we want to immediately use `runPCA()` instead of including the `calculatePCA()` step? I can see why I might want to know about `calculatePCA()` but I would think that we want to encourage participants to add to their SingleCellExperiment.
* If we're concerned about time, we could think about dropping the t-SNE comparison. (Or kicking that comparison to an exercise...)

Because the flow looked good overall, I did go ahead and leave more detailed comments about wording, etc.

jashapiro commented:
I think I addressed all of @jaclyn-taroni's comments, but I also made a more substantive changes since the review:

I wanted to shorten the normalization a bit to keep day 2 (scRNAseq day 1) from getting too long, so I moved the discussion of modeling high variance genes to this notebook. The resulting relationship between mean expression and variance is a bit funny looking, b/c there is one kind of outlying gene with high expression, but I decided to leave it. 

I also made some more decisions about which chunks should be live, as I hadn't yet done that.

I'm not too worried about the length of this notebook, as a good chunk of it is review/extension of the previous day, which I think will work well. Trying to keep the rest not too long.

jaclyn-taroni commented on scRNA-seq/04-dimension_reduction_scRNA.Rmd:
```diff
@@ -31,226 +42,294 @@ library(magrittr)
-detected <- rowData(tm_sce_filtered)$detected > 5
-expressed <- rowData(tm_sce_filtered)$mean > 0.1
+filtered_sce <- scater::addPerFeatureQC(filtered_sce)
+detected <- rowData(filtered_sce)$detected > 5
```
Hm, I wouldn't necessarily expect this to do what the text says (based on the name) but okay!

jashapiro replied:
You mean you wouldn't expect `detected` to be a percent measure? Yeah, neither would I, but that is what it seems to be: 
> detected: numeric, the percentage of observations above threshold. 
https://rdrr.io/bioc/scuttle/man/perFeatureQCMetrics.html

jaclyn-taroni commented on scRNA-seq/04-dimension_reduction_scRNA.Rmd:
```diff
@@ -31,226 +42,294 @@ library(magrittr)
-dim(tm_normalized) # column count is the number of cells
+The variation in gene expression we see among cells comes from a combination of variation due to technical effects and the biology we really care about. 
+In order to roughly account for this we could just take the largest variance genes, on the assumption that low variance genes are mostly just noise. 
+This is the default approach that `runPCA()` and `calculatePCA()` take, using 500 genes with the 
+greatest variance across cells to calculate the PCA matrix. 
```
```suggestion
This is the default approach that `runPCA()` and `calculatePCA()` take, using 500 genes with the greatest variance across cells to calculate the PCA matrix. 
```

jaclyn-taroni commented on scRNA-seq/04-dimension_reduction_scRNA.Rmd:
````diff
@@ -31,226 +42,294 @@ library(magrittr)
+```
 
-In the chunk below, we calculate the PCs and plot the first two, colored by tissue, much as we did before.
+Now we can use `scran::getTopHVGs()` to select the genes that have the most biological variation (according to the model) and recalculate PCA scores using only those genes. 
````
I feel pretty certain that we've already had this conversation, but can you remind me (roughly) how genes with biological variation are defined by this particular method? I can understand why you might not want to get into it here (again, we've probably had this exact conversation before!) but it is helpful for instructors to understand.

jashapiro replied:
I _believe_ (I checked, [I seem to be correct](https://rdrr.io/bioc/scran/man/modelGeneVar.html)) it models the gene variance relationship (shown in the plot) with a loess model or similar, then uses the residuals as the measure of biological variation. 

```suggestion
Now we can use `scran::getTopHVGs()` to select the genes that have the most biological variation (according to the model) and recalculate PCA scores using only those genes. 
(In practice, we are selecting the genes with the largest residual variation after removing technical variation modeled by the mean/variance relationship.)
```

Something like that?

jaclyn-taroni commented on scRNA-seq/data/hodgkins/hs_mitochondrial_genes.tsv:
```diff
@@ -0,0 +1,38 @@
+gene_id	gene_name	seqnames	start	end	width	strand	gene_biotype	seq_coord_system	description	gene_id_version	symbol	entrezid
```
Wanted to just check in about the decision to include this in version control and how that might impact `skel` prep.

jashapiro replied:
I don't think it will affect skel prep, but yeah, I don't know. I figured it was pretty small and the mouse file was already in there, but I felt _bad_ when I added it.

jaclyn-taroni replied:
A better question I could have asked instead: Should this be in S3 instead of tracked here?

jashapiro replied:
Since the mouse version file just came up as an issue for exercise notebooks, the answer to this is yes. I filed a separate issue: https://github.com/AlexsLemonade/training-modules/issues/472 to update the scripts (and remove this version).

jaclyn-taroni commented on scRNA-seq/04-dimension_reduction_scRNA.Rmd:
````diff
@@ -31,226 +42,294 @@ library(magrittr)
+These NAs can be a problem for filtering with a Boolean vector, as we did above, so instead we will use the `which()` function to get the *positions* of the cells that pass our filter and select the columns we want using that.
+
+```{r filter_empty, live = TRUE}
+cells_to_retain <- which(droplet_stats$FDR <= 0.01)
````
Comment for when it gets presented: would make sure to take a look at `cells_to_retain` the global environment

jaclyn-taroni reviewed (approved):
Looks good to me! Couple questions/points for discussion, but nothing that I think should hold this up.
